cf-keycrypt/cf-keycrypt.c

```diff
@@ -0,0 +1,504 @@
+        found = Address2Hostkey(hash, sizeof(hash), ipaddress);
+    }
+    if (found)
+    {
+        snprintf(buffer, BUFSIZE * sizeof(char), "%s/ppkeys/root-%s.pub", WORKDIR, hash);
+        freeaddrinfo(result);
+        return buffer;
+    }
+    else
+    {
+        for(res = result; res != NULL; res = res->ai_next)
+        {
+            inet_ntop(
+                res->ai_family,
+                get_in_addr((struct sockaddr *)res->ai_addr),
+                ipaddress,
+                sizeof(ipaddress));
+            snprintf(
+                buffer, BUFSIZE * sizeof(char), "%s/ppkeys/root-%s.pub",
+                WORKDIR, ipaddress);
+            if (file_exist(buffer))
+            {
+                freeaddrinfo(result);
+                return buffer;
+            }
+        }
+    }
+    return NULL;
+}
+
+RSA *readseckey(const char *privkey_path)
+{
+    FILE *fp = fopen(privkey_path,"r");
+
+    if (fp == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not open private key '%s'", privkey_path);
+        return NULL;
+    }
+    RSA *privkey = PEM_read_RSAPrivateKey(fp, (RSA **)NULL, NULL,
+                                          (void *) passphrase);
+    if (privkey == NULL)
+    {
+        unsigned long err = ERR_get_error();
+        Log(LOG_LEVEL_ERR, "Could not read private key '%s': %s",
+            privkey_path, ERR_reason_error_string(err));
+    }
+    fclose(fp);
+    return privkey;
+}
+
+RSA *readpubkey(const char *pubkey_path)
+{
+    FILE *fp = fopen(pubkey_path, "r");
+
+    if (fp == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not open public key '%s'", pubkey_path);
+        return NULL;
+    }
+
+    RSA *pubkey = PEM_read_RSAPublicKey(
+        fp, NULL, NULL, (void *)passphrase);
+    if (pubkey == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not read public key '%s'", pubkey_path);
+    }
+    fclose(fp);
+    return pubkey;
+}
+
+long int rsa_encrypt(
+    const char *pubkey_path, const char *input_path, const char *output_path)
+{
+    int ks = 0;
+    unsigned long len = 0, ciphsz = 0;
+
+    RSA* pubkey = readpubkey(pubkey_path);
+    if (pubkey == NULL)
+    {
+        return -1;
+    }
+
+    FILE *input_file = fopen(input_path, "r");
+    if (input_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not open input file '%s'", input_path);
+        RSA_free(pubkey);
+        return -1;
+    }
+
+    FILE *output_file = fopen(output_path, "w");
+    if (output_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Could not create output file '%s'", output_path);
+        fclose(input_file);
+        RSA_free(pubkey);
+        return -1;
+    }
+
+    ks = RSA_size(pubkey);
+    char tmp_ciphertext[ks], tmp_plaintext[ks];
+
+    srand(time(NULL));
+    bool error_return = false;
+    while (!feof(input_file))
+    {
+        memset(tmp_plaintext, '\0', ks);
+        memset(tmp_ciphertext, '\0', ks);
+        len = fread(
+            tmp_plaintext, 1, ks - RSA_PKCS1_PADDING_SIZE, input_file);
```

Comment:
So it keeps reading from a file 245 byte chunks, and RSA-encrypts each and every one separately using PKCS #1 v1.5 padding, and appends to a file, and each block is fully independent from the other without any chaining whatsoever. This is definitely [fishy](https://stackoverflow.com/questions/13908407/openssl-rsa-public-encrypt-returned-size-always-rsa-sizersa) but I am not sure if there are specific vulnerabilities. Maybe @estenberg can comment ?

Reply:
@jimis you are right, this is not good because 1) it leaks information as identical input chunks result in identical encrypted data 2) encrypting large volumes with a public key can leak information about the private key and 3) it is very inefficient (compared to symmetric crypto). At the very least we need to set a hard limit on the size (e.g. you can only encrypt 245 bytes or less), so it would really only work for password strings and the like.

The "right way" to do this is of course to encrypt a random symmetric key with RSA and then use that key with a block or stream cipher in a secure mote (e.g. CBC, CTR, etc.)